cranelift-codegen/src/redundant_reload_remover.rs

```diff
@@ -0,0 +1,462 @@
+    cur: EncCursor<'a>,
+
+    // Cached ISA information.  We save it here to avoid frequent virtual function calls on the
+    // `TargetIsa` trait object.
+    reginfo: RegInfo,
+
+    // References to contextual data structures we need.
+    domtree: &'a DominatorTree,
+
+    // The running state
+    state: &'a mut RedundantReloadRemover,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        self.cleanup_func();
+    }
+}
+
+// ======== A description of the redundant-fill-removal algorithm ========
+//
+// The simplest incarnation simply detects fills from the same stack slot into the same register,
+// and removes them:
+//
+//  v1 = fill v0
+//  ...
+//  v2 = fill v0
+//
+// The second fill is redundant provided that
+//   (1) v1 and v2 have the same register, and
+//   (2) "..." overwrites neither that register nor v0's stack slot, and
+//   (3) the second fill is dominated by the first.
+// Condition (3) always holds since we're doing this only inside EBBs, but we have to check (1)
+// and (2).
+//
+// In this example, the instruction processing loop below creates an environment entry on
+// encountering "v1 = fill v0", and uses that entry to determine that "v2 = fill v0" is redundant.
+//
+// The loop must also, even in the simplest version, check all instructions for writes to
+// registers and stack slots, and invalidate environment entries accordingly.
+//
+// === First refinement -- implemented
+//
+// The simplest version only observes that a register and stack slot have the same value when it
+// finds a fill.  However, a spill also causes a register and stack slot to have the same value.
+// Hence this first refinement is to also record such spill-created equivalences in the
+// environment, in the hope of increasing our chances of proving later fills redundant.
+//
+// === Second refinement -- implemented
+//
+// For a fill, even if we can't find an exact match in the environment, finding a load of the same
+// spill slot into a different register is still a win, since we can convert the fill into a
+// reg-reg copy.
+//
+// === Other tricks that don't seem to help
+//
+// Tracking the effects of InstructionData::Copy in the environment for the case when both source
+// and destination are registers.  But that didn't cause any more fills to be removed or converted
+// to copies.  It's not clear why.
+//
+// It's also possible for a spill to be detected as redundant, if such a binding already appears
+// in the environment.  This was also implemented, but it was almost entirely ineffective, so was
+// removed again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA value name associated
+// with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offset: i32,
+    size: u32,
+}
+
+// Find out if the range `[offset, +size)` overlaps with the range in `si`.
+fn any_overlap(si: &SlotInfo, offset: i32, size: u32) -> bool {
```

Comment:
nit: since it's a boolean testing function, I'd prefer it to be either starting with "has" or "is" or be a verb at the singular form: what do you think about `overlaps`?